artifactory/commands/terraform/terraformpublish.go

```diff
@@ -0,0 +1,365 @@
+	go func() {
+		defer producer.Done()
+		pwd, err := os.Getwd()
+		if err != nil {
+			log.Error(err)
+			errorsQueue.AddError(err)
+		}
+		// Walk and upload directories which contain '.tf' files.
+		err = tpc.walkDirAndUploadTerraformModules(pwd, producer, errorsQueue, uploadSummary, addTaskWithError)
+		if err != nil && err != io.EOF {
+			log.Error(err)
+			errorsQueue.AddError(err)
+		}
+	}()
+}
+
+// ProduceTaskFunk Provided as a parameter to 'walkDirAndUploadTerraformModules' function for testing reasons.
+type ProduceTaskFunk func(producer parallel.Runner, uploadService *services.UploadService, uploadSummary *specutils.Result, target string, archiveData *services.ArchiveUploadData, errorsQueue *clientutils.ErrorsQueue) (int, error)
+
+func addTaskWithError(producer parallel.Runner, uploadService *services.UploadService, uploadSummary *specutils.Result, target string, archiveData *services.ArchiveUploadData, errorsQueue *clientutils.ErrorsQueue) (int, error) {
+	return producer.AddTaskWithError(uploadService.CreateUploadAsZipFunc(uploadSummary, target, archiveData, errorsQueue), errorsQueue.AddError)
+}
+
+func (tpc *TerraformPublishCommand) walkDirAndUploadTerraformModules(pwd string, producer parallel.Runner, errorsQueue *clientutils.ErrorsQueue, uploadSummary *specutils.Result, produceTaskFunk ProduceTaskFunk) error {
+	return filepath.WalkDir(pwd, func(path string, info fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		pathInfo, e := os.Lstat(path)
+		if e != nil {
+			return e
+		}
+		// Skip files and check only directories.
+		if !pathInfo.IsDir() {
+			return nil
+		}
+		isTerraformModule, e := checkIfTerraformModule(path)
+		if e != nil {
+			return e
+		}
+		if isTerraformModule {
+			uploadParams := tpc.uploadParamsForTerraformPublish(pathInfo.Name(), strings.TrimPrefix(path, pwd+string(filepath.Separator)))
+			archiveData := createTerraformArchiveUploadData(uploadParams, errorsQueue)
+			dataHandlerFunc := GetSaveTaskInContentWriterFunc(archiveData, *uploadParams, errorsQueue)
+			// Collect files that matches uploadParams and write them in archiveData's writer.
+			e = services.CollectFilesForUpload(*uploadParams, nil, nil, dataHandlerFunc)
+			if e != nil {
+				log.Error(e)
+				errorsQueue.AddError(e)
+			}
+			e = archiveData.GetWriter().Close()
+			if e != nil {
+				log.Error(e)
+				errorsQueue.AddError(e)
+			}
+			// In case all files were excluded and no files were written to writer- skip this module.
+			if archiveData.GetWriter().IsEmpty() {
+				return filepath.SkipDir
+			}
+			uploadService := createUploadServiceManager(tpc.serverDetails, errorsQueue)
+			_, e = produceTaskFunk(producer, uploadService, uploadSummary, uploadParams.Target, archiveData, errorsQueue)
+			if e != nil {
+				log.Error(e)
+				errorsQueue.AddError(e)
+			}
+			// SkipDir will not stop the walk, but will jump to the next directory.
```

Comment:
...but will jump to the next directory. --> ...but it will make us jump to the next directory.